### PR TITLE
rpc: add raft connection class

### LINKED ISSUE #111262: kvserver: separate Raft RPC connection class
Currently, most RPC traffic is sent over a single TCP connection (the `default` class). Unrelated streams on this connection can be subject to head-of-line blocking under heavy load, often due to RPC processing latency rather than TCP processing.

We should consider splitting out a separate connection class for Raft traffic, and potentially one for snapshot traffic as well, to reduce interference with foreground RPC traffic. This might also to a small extent mitigate bandwidth-delay product impact on high-latency links (see #111241) and packet loss (see #110099).

Jira issue: CRDB-31842

Epic CRDB-32846

### PR DESCRIPTION
**Pull request by pav-kv**
Resolves #111262

Release note (performance improvement): this change introduces a separate RPC connection class for most raft traffic. This improves isolation and reduces interference with the foreground SQL traffic, which reduces chances of head-of-line blocking caused by unrelated traffic under high load conditions. The new `COCKROACH_RAFT_USE_DEFAULT_CONNECTION_CLASS` env variable can be set to use the default connection class instead (the previous behaviour).

### CODE REVIEW

File: pkg/kv/kvserver/replica_init.go [hunk 1]
```diff
@@ -396,7 +396,7 @@ func (r *Replica) setDescLockedRaftMuLocked(ctx context.Context, desc *roachpb.R
 
 	r.rangeStr.store(r.replicaID, desc)
 	r.isInitialized.Set(desc.IsInitialized())
-	r.connectionClass.set(rpc.ConnectionClassForKey(desc.StartKey))
+	r.connectionClass.set(rpc.ConnectionClassForKey(desc.StartKey, rpc.RaftClass))
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica_init.go:399]: @erikgrinaker Looks good? Should we add a cluster setting to flip this back to `DefaultClass` if needed?

Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_init.go:399]: Yeah, seems fine. We could consider something like `rpc.ShouldUseSystemConnectionClass()` instead, but I guess it's just going to be more verbose.

>Should we add a cluster setting to flip this back to DefaultClass if needed?

Nah, I doubt this is going to cause any problems. We can add an envvar if we think we need one, but I don't see why we would.

Comment (pav-kv) [reply pkg/kv/kvserver/replica_init.go:399]: > We could consider something like rpc.ShouldUseSystemConnectionClass() instead, but I guess it's just going to be more verbose.

This seems ok, as we already have [some code](https://github.com/cockroachdb/cockroach/blob/9530d1adb5ae5c6136455178617c56201deb443b/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L959-L964) for figuring out the class outside the `rpc` package. So could have more.

I feel like `rpc.ConnectionClassForKey` is a bit out of place. Seemingly, the `rpc` package shouldn't be aware of the keys layout. So, arguably, this code should move to upper layers anyway, and just tell the `rpc` package which class to use.

Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_init.go:399]: 🤷 I could see us having a generalized mechanism for marking system ranges, but it isn't clear that we'd want all system ranges to use the system class -- for example, consider the time series ranges or the `system` database (which contains a ton of observability data).

It seems fine for now to keep something like this in the `rpc` package, and limit the scope specifically to RPC class conditions, but I don't have a strong opinion on how to structure it. We can maybe revisit this once we've implemented all the classes and have a better sense of what's needed.

Comment (pav-kv) [reply pkg/kv/kvserver/replica_init.go]: Keeping this simple for now. Added an env var to have a backstop. Made it static to a) avoid the overhead of reading the env var in every `Replica` init path, b) avoid the "eventually consistent" process state in which we use both `DefaultClass` and `RaftClass` while switching between the two, and have to garbage collect one.

Let's get back to system ranges aspects in #116958.

**Comment by pav-kv** [comment]
@erikgrinaker ready for review

**Review by erikgrinaker — APPROVED** [review]
LGTM.

We considered also using a separate class for Raft snapshots, but that's probably overkill given the snapshot queue only sends one at a time.

**Comment by pav-kv** [comment]
bors r=erikgrinaker

**Pull request merged**